lib/Dialect/D2M/Transforms/InsertDstRegisterGC.cpp

```diff
@@ -0,0 +1,1257 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/D2M/Analysis/DstAnalysis.h"
+#include "ttmlir/Dialect/D2M/Analysis/DstAnalysisGraphColoring.h"
+#include "ttmlir/Dialect/D2M/Analysis/DstCapacityAnalysis.h"
+#include "ttmlir/Dialect/D2M/IR/D2MOps.h"
+#include "ttmlir/Dialect/D2M/Transforms/GraphColoringStrategy.h"
+#include "ttmlir/Dialect/D2M/Transforms/Passes.h"
+#include "ttmlir/Dialect/D2M/Utils/TileMatmulUtils.h"
+
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/Dialect/Affine/Analysis/LoopAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/IR/PatternMatch.h"
+#include "ttmlir/Utils.h"
+
+#include <algorithm>
+#include <numeric>
+
+namespace mlir::tt::d2m {
+#define GEN_PASS_DEF_D2MINSERTDSTREGISTERGC
+#include "ttmlir/Dialect/D2M/Transforms/Passes.h.inc"
+
+namespace {
+
+// Loop context information for DST index computation.
+// Tracks the loop nest containing a DST access and its iteration space.
+struct LoopContext {
+  llvm::SmallVector<mlir::affine::AffineForOp, 4>
+      loopNest;                             // Outermost to innermost.
+  llvm::SmallVector<int64_t, 4> tripCounts; // Trip count for each loop.
+  int64_t totalIterations;                  // Product of all trip counts.
+
+  LoopContext() : totalIterations(1) {}
```

Comment:
:warning: **cppcoreguidelines\-use\-default\-member\-init** :warning:
use default member initializer for `` totalIterations ``
```suggestion
  LoopContext() : {}
```